POS Bookings — M1 Foundation (Feature Flag, List, Detail, Service Layer) (#16635)
woocommerce/woocommerce-ios: Modules/Sources/PointOfSale/Presentation/Bookings/BookingsSpec.md

### Milestone 1 — Core Flow & Payment (Feb 13)
**Goal:** End-to-end payment for a booking with minimal scope.

**Navigation & Entry**
- "Bookings" button in the POS floating menu — **only visible when the merchant's site supports bookings** (has the WooCommerce Bookings plugin active with bookable products or existing bookings). This mirrors the main app's Bookings tab eligibility check (`BookingsTabEligibilityChecker`).
- Bookings opens as a full-screen overlay on top of POS

**Booking List (left pane)**
- Shows today's bookings, paginated (load more on scroll)
- Each row displays: customer name, booking amount, service name, time, payment status
- Search bookings by customer name
- Pull-to-refresh
- Empty state when no bookings exist
- Loading state with ghost/skeleton rows
- Error state with retry
- In-memory caching: retains content when bookings is closed and reopened

**Booking Detail (right pane)**
- Selecting a booking shows its details: service, date/time, customer, resource (staff), amount
- Payment action buttons: "Pay by Card" and "Pay by Cash"

> - Buttons hidden when booking is already paid or cancelled

Reviewer: ```suggestion
- Payment action button: "Collect Payment"
- Button hidden when booking is already paid or cancelled
```

#### A4. POSBookingsContainerView — split view
**New file:** `PointOfSale/Presentation/Bookings/POSBookingsContainerView.swift`

- `CustomNavigationSplitView` (from `POSOrdersView`)
- Error/empty/content states matching `POSOrdersView` pattern
- Auto-select first booking on iPad

**Mirrors:** `PointOfSale/Presentation/Orders/POSOrdersView.swift`

#### A5. POSBookingListView — pagination and search
**New file:** `PointOfSale/Presentation/Bookings/POSBookingListView.swift`

- "Bookings" header + close button + search icon
- `InfiniteScrollView` for pagination
- `POSSearchField` for search
- Ghost loading rows, pull-to-refresh, inline errors

**Mirrors:** `PointOfSale/Presentation/Orders/POSOrderListView.swift`

#### A6. POSBookingRowView
**New file:** `PointOfSale/Presentation/Bookings/POSBookingRowView.swift`

- `ScaledMetric` + `dynamicTypeSize` for accessibility
- `posItemCardBorderStyles()` for selection highlight
- 3-line layout: customer+amount, service+time, status badge

#### A7. POSBookingDetailView
**New file:** `PointOfSale/Presentation/Bookings/POSBookingDetailView.swift`

- Sections: booking info, status badges, customer, payment breakdown, action buttons

> - Contextual states: paid → checkmark, cancelled → label, noLinkedOrder → explanation

Reviewer: ```suggestion
- Contextual states: paid → checkmark, cancelled → label, noLinkedOrder → explanation, unpaid with order → collect payment button
```

#### A8. Empty/Loading/Error views
**New files:** `POSBookingDetailsEmptyView.swift`, `POSBookingDetailsLoadingView.swift`

#### A9. Entry point wiring
**Modify:** `PointOfSale/Presentation/PointOfSaleEntryPointView.swift`

Pass strategy factory to booking list controller. `POSFloatingControlView` already has bookings button behind `.pointOfSaleBookings`.

---

### Stream B: Payment & Checkout

#### B1. POSBookingPaymentController — full payment state machine

> **New file:** `PointOfSale/Controllers/POSBookingPaymentController.swift`

Reviewer: If this can be shared, i.e. a full extraction of the payment wrangling code, that would be good. But I'm not making it a requirement in the plan, maybe something to investigate afterwards.

Author: Yes. I agree that it should be decided when diving deeper in the payment reuse. I doubt we can one-shot it with a good spec prompt, given many details that could affect the decision.

Here's an additional investigation that I run with suggestion:

**Option A (Extract POSPaymentOrchestrator):** 

Pull out a reusable @Observable class from PointOfSaleAggregateModel that owns everything payment-related — the 3 Combine subscription chains (paymentEvent → alertViewModel, paymentEvent → inlineMessage, paymentEvent → cardPaymentState), the reader connection/reconnection
  logic, the cash payment flow, and cancelThenCollectPayment. This orchestrator would take two abstractions: an "order provider" closure (cart creates orders, bookings fetch existing ones) and an "action configuration" (what happens on success, what the error buttons do). PointOfSaleAggregateModel would delegate to
  this orchestrator internally, and POSBookingsModel would create its own instance with booking-specific closures. Maximum shared code, but requires refactoring the aggregate model which is risky since it's the central POS coordinator.

 **Option B (Reuse views only, duplicate orchestration):** 

Leave PointOfSaleAggregateModel completely untouched. Create a separate POSBookingPaymentController that reimplements the same orchestration logic (subscriptions, reconnection, cash flow) but sourcing orders from the booking's linked order instead of creating
  them. Extract the private payment sub-views from TotalsView (CardPaymentView, CashPaymentView, PaymentViewContent, CashPaymentButton) into shared internal views so both TotalsView and the booking payment view can render the same UI. This is the safest approach — zero risk to existing functionality — but the ~120
  lines of orchestration logic are duplicated and could diverge over time.

 **Option C (Hybrid — extract subscriptions + views, thin wrappers):** 

Extract the 3 Combine subscription chains into a shared helper (e.g. POSPaymentEventSubscriber) that any controller can use, and extract the private sub-views as in Option B. Each context (cart/bookings) keeps its own thin controller that provides
  the order and wires the context-specific closures, but delegates the repetitive subscription plumbing to the shared helper. Less invasive than A (aggregate model only changes its internal implementation to use the helper), less duplication than B (the subscription chains aren't copied)

----

I suspect we could take Option B just because duplication is not as much, and we would be risking less breaking anything in the main flow. However, Option A could be tried just to see how many changes it generates.

Reviewer 2: > I suspect we could take Option B just because duplication is not as much, and we would be risking less breaking anything in the main flow. However, Option A could be tried just to see how many changes it generates.

Since we're already taking shortcuts willingly in other places, I'd opt as well for duplicating the orchestration if that means a safer approach for the rest of the POS. It should also be easier to do B -> A later on before they diverge, and seems less risky than attempt A directly with the timeline.

```swift
@MainActor @Observable
final class POSBookingPaymentController {
    private(set) var paymentState: PointOfSalePaymentState
    var cardPresentPaymentAlertViewModel: PointOfSaleCardPresentPaymentAlertType?
    private(set) var cardPresentPaymentInlineMessage: PointOfSaleCardPresentPaymentMessageType?
    private(set) var cardReaderConnectionStatus: CardPresentPaymentReaderConnectionStatus
    let booking: POSBooking
}
```

**Key methods (mirror AggregateModel):**
- `startPaymentWhenCardReaderConnected()` — subscription-based: if connected, collect immediately; otherwise subscribe to connection status and collect on connect
- `collectCardPayment()` — fetch order via `orderProvider.fetchOrder(siteID:orderID:)`, then `cardPaymentFacade.collectPayment(for:using:channel:)`
- `cancelThenCollectPayment()` — cancel current and retry
- `publishPaymentMessages()` — 3 subscription chains:
  1. Events → `cardPresentPaymentAlertViewModel` (modals)
  2. Events → `cardPresentPaymentInlineMessage` (inline)
  3. Events → `paymentState.card` via `PointOfSaleCardPaymentState(from:using:)`

> - On `cardPaymentSuccessful`: call `bookingService.markBookingAsPaid()`

Reviewer: ```suggestion
- On `cardPaymentSuccessful`: call `bookingService.markBookingAsPaid()`
- Even though the order exists, we still need `validatingOrder` phase, as the card payment service will check it is payable.
```

**Key differences from AggregateModel:**

> - No `validatingOrder` phase (order pre-exists)

Reviewer: ```suggestion
```

- `validatingOrderError` CAN happen (order fetch failure)
- No order creation, no cart, no order stage

> - Error action closures reset to idle (no "new order" or "edit order" concepts)

Reviewer: ```suggestion
- Error action closures reset to idle and go back to the booking (no "new order" or "edit order" concepts.) We need to make the button text modifiable for the context.
```

**Cash methods:**
- `startCashPayment()` — cancel card, set `cash = .collectingCash`
- `cancelCashPayment()` — set `cash = .idle`, resume card if connected
- `collectCashPayment(changeDueAmount:)` — mark paid, set `cash = .paymentSuccess`

**Reader reconnection:** subscription-based (like AggregateModel)

**Reuses:**
- `PointOfSaleCardPaymentState(from:using:)` initializer
- `PointOfSaleCardPresentPaymentEventPresentationStyle`
- `CardPresentPaymentFacade`
- Subscription patterns from `PointOfSaleAggregateModel` lines 352-654

#### B2. POSBookingPaymentView — card payment UI
**New file:** `PointOfSale/Presentation/Bookings/POSBookingPaymentView.swift`

State-driven rendering based on `paymentState.card`:
- `idle/preparingReader/acceptingCard/cardInserted` — inline messages + booking amount
- `processingPayment` — purple background
- `paymentError` — error with retry
- `cardPaymentSuccessful` — "Email Receipt" + "Done"

> Uses: `PointOfSaleCardPresentPaymentInLineMessage`, `PointOfSaleCardPresentPaymentAlert`, `PointOfSaleCardPresentPaymentReaderDisconnectedMessageView`

Reviewer: To explore deeper with Claude – does this need to be booking specific, or can it be extracted from the existing code, especially if it can be done at a slightly higher level so it handles the switching between card and cash.

Author: Investigation from Claude:

## What is Currently Order-Specific vs Generic in the Payment Code

### Already generic (context-agnostic):

1. **`CardPresentPaymentFacade`** — the protocol takes an `Order` and collects payment. It knows nothing about carts, bookings, or order creation. Its `collectPayment(for:using:channel:)` accepts a plain `Order` struct.

2. **`PointOfSalePaymentState` / `PointOfSaleCardPaymentState` / `PointOfSaleCashPaymentState`** — pure state enums. They track card states (`idle`, `acceptingCard`, `processingPayment`, `paymentError`, `cardPaymentSuccessful`, etc.) and cash states (`idle`, `collectingCash`, `paymentSuccess`). No order/cart/booking references.

3. **`PointOfSaleCardPaymentState(from:using:)` initializer** — maps `CardPresentPaymentEvent` to card payment state. Completely generic.

4. **All card present payment message/alert views** — `PointOfSaleCardPresentPaymentInLineMessage`, `PointOfSaleCardPresentPaymentErrorMessageView`, processing/success/tap-swipe views, etc. These are driven by view models and closures, not by `PointOfSaleAggregateModel` directly.

5. **`CollectCashViewHelper`** — currency formatting, change due calculation. Takes `orderTotal` as a string. Fully generic.

6. **`PointOfSaleCardPresentPaymentEventPresentationStyle`** — maps `CardPresentPaymentEventDetails` to presentation style (alert vs message). This is context-agnostic *except* for its `Dependencies` struct (see below).

### Order-specific (would need abstraction for bookings):

1. **`PointOfSaleAggregateModel`** — the ~700-line god model that owns cart, order controller, payment state, card reader subscriptions, and payment message publishing. It is *the* coupling point. All payment logic (lines 352-654) lives here, tightly interwoven with cart/order concepts.

2. **`PointOfSaleCardPresentPaymentEventPresentationStyle.Dependencies`** — this is where the order-specific button labels/actions live:
   - `paymentCaptureErrorNewOrderAction` calls `startNewCart()`, button says **"New order"**
   - `paymentIntentCreationErrorEditOrderAction` calls `addMoreToCart()`, button says **"Edit order"**
   - `tryPaymentAgainBackToCheckoutAction` calls `cancelThenCollectPayment()`
   - The `Dependencies` struct itself is context-agnostic (just closures and strings), but the *values* injected are order-specific.

3. **`PointOfSalePaymentSuccessView`** — directly uses `@Environment(PointOfSaleAggregateModel.self)` to call `posModel.sendReceipt(to:)` and `posModel.startNewCart()`.

4. **`PaymentsActionButtons`** — uses `@Environment(PointOfSaleAggregateModel.self)` for "New order" button calling `posModel.startNewCart()`.

5. **`PointOfSaleCollectCashView`** — uses `@Environment(PointOfSaleAggregateModel.self)` for `posModel.cancelCashPayment()`, `posModel.collectCashPayment(changeDueAmount:)`.

6. **`PointOfSaleCardPresentPaymentCaptureErrorMessageViewModel`** — hardcoded "New order" button label.

7. **`PointOfSaleCardPresentPaymentIntentCreationErrorMessageViewModel`** — hardcoded "Edit order" button label.

8. **`collectCardPayment()`** in AggregateModel — accesses `internalOrderState` to get the `Order` for payment.

## Is Extraction Feasible?

**Yes, and the architecture is already partially set up for it.** The majority of the payment *mechanism* is generic. The coupling is concentrated in a few specific areas:
- The `@Environment(PointOfSaleAggregateModel.self)` usage in ~5 views
- The `Dependencies` struct closures/labels in the presentation style
- The order-fetching logic (`collectCardPayment()` pulling from `internalOrderState`)

## Options

### Option A: Extract a `POSPaymentController` protocol/class (recommended)

Extract the payment orchestration from `PointOfSaleAggregateModel` lines 352-654 into a standalone `POSPaymentController` that both `PointOfSaleAggregateModel` and `POSBookingsModel` can own.

**What it would own:**
- `paymentState: PointOfSalePaymentState`
- `cardPresentPaymentAlertViewModel`
- `cardPresentPaymentInlineMessage`
- `cardReaderConnectionStatus` / `cardReaderUpdateState`
- All Combine subscriptions for payment events
- `publishPaymentMessages()`, `publishCardReaderConnectionStatus()`, `publishCardReaderUpdateState()`
- `startPaymentWhenCardReaderConnected()`, `collectPayment(for:)`, `cancelThenCollectPayment()`
- `startCashPayment()`, `cancelCashPayment()`, `collectCashPayment(changeDueAmount:)`
- Reader reconnection observation

**What it would take as dependencies (via init or protocol):**
- `CardPresentPaymentFacade` (already a protocol)
- An `orderProvider` closure/protocol: `() -> Order?` — AggregateModel provides this from its order controller; Bookings provides this by fetching the booking linked order
- A `PaymentActionContext` struct with configurable button labels and actions for error recovery:
  - `captureErrorPrimaryAction` / `captureErrorSecondaryAction` (replaces hardcoded "New order")
  - `intentCreationErrorSecondaryAction` (replaces hardcoded "Edit order")
  - `onPaymentSuccess` callback
- `cashPaymentHandler`: a closure/protocol to execute the actual cash payment recording (order controller for orders, booking service for bookings)

**Pros:**
- Single source of truth for payment state machine logic (~300 lines of non-trivial Combine code)
- Both contexts get identical reader connection handling, error recovery, and reconnection patterns
- Existing views (`PointOfSaleCardPresentPaymentInLineMessage`, error views) work unchanged — they already consume view models, not the model directly
- The 5 views using `@Environment(PointOfSaleAggregateModel.self)` would need small changes (inject closures instead or use a protocol environment)
- Bug fixes in payment flow automatically apply to both contexts

**Cons:**
- Requires refactoring `PointOfSaleAggregateModel` to delegate to the new controller (medium effort)
- The few views that use `@Environment(PointOfSaleAggregateModel.self)` (`PaymentSuccessView`, `PaymentsActionButtons`, `CollectCashView`) need to be parameterized with closures or a protocol
- Testing the extracted controller requires verifying no regressions in the existing POS flow

### Option B: Duplicate the payment logic in `POSBookingPaymentController` (current spec approach)

The BookingsSpec at B1 proposes a new `POSBookingPaymentController` that "mirrors" the AggregateModel payment methods but is booking-specific.

**Pros:**
- No risk to existing POS payment flow
- Faster to implement initially
- Can diverge if booking payment needs differ

**Cons:**
- ~300 lines of duplicated Combine subscription and state machine code
- Two places to fix bugs in payment flow
- The BookingsSpec already acknowledges reusing all the same types (`PointOfSaleCardPaymentState`, `CardPresentPaymentFacade`, presentation styles) — the duplication is *only* in the orchestration glue

### Option C: Make `PointOfSaleAggregateModel` itself generic enough

Inject an order provider and customize the error actions, keeping payment logic in the aggregate model.

**Pros:** Minimal code changes
**Cons:** The aggregate model is already ~700 lines and manages cart, items, barcode scanning, search, coupons, catalog sync. Adding booking awareness would make it harder to maintain. Not really feasible without making it worse.

## Recommendation

**Option A is the right path.** The payment orchestration code in `PointOfSaleAggregateModel` (lines 352-654) is already functionally context-agnostic — it just needs to be *structurally* separated. The coupling points are well-defined and limited:

1. **Order provider** — trivially abstractable (closure returning `Order?`)
2. **Error button labels/actions** — already parameterized via `Dependencies` closures, just needs the hardcoded strings in 2 view models to become configurable
3. **Post-success actions** — "New order" vs "Done/Back to booking" — just a callback
4. **Cash payment recording** — already a single method call to order controller, replaceable with a closure
5. **5 views using `@Environment`** — switch to closures or a protocol-based environment (the spec already proposes this for `POSBookingCashPaymentView`: "Closure-based actions, no `@Environment(PointOfSaleAggregateModel.self)`")

The fact that the BookingsSpec already plans to reuse `PointOfSalePaymentState`, `PointOfSaleCardPaymentState(from:using:)`, `CardPresentPaymentFacade`, and all the presentation style code confirms that the payment mechanism is generic — it is only the orchestration glue that is currently embedded in the aggregate model.

**Suggested approach:** Extract `POSPaymentController` first (with the AggregateModel delegating to it), verify no regressions in existing POS, then have `POSBookingsModel` use the same controller with booking-specific configuration.

Author: It can be done in multiple ways. Claude after analysis, also sees a way to make `TotalsView` itself more configurable with injected dependencies to reuse straightaway. I wonder if we should be less prescriptive in this spec. Payments are complicated, I'd much rather have developer more freedom when we start developing it. Spec could contain general guidelines that tries to reuse as much of the existing payment flows as possible while keeping the balance of not breaking the existing flows.

Author: [Claude] Investigated this — it can largely be extracted from existing code. TotalsView and its sub-views are almost entirely payment-generic. Only 4 coupling points are cart-specific:

1. **Success view** (`PointOfSalePaymentSuccessView`) — reads `posModel` from environment, "New order" / barcode scanning
2. **Cash collection** (`PointOfSaleCollectCashView`) — reads `posModel` from environment for cancel/collect
3. **Error button text** — "New order" / "Edit order" / "Go back to checkout" hardcoded in 3 error VMs
4. **Discount field** — `TotalsFieldsContent` takes a `Cart` to check coupons

Everything else (card reader UI, inline messages, `TotalsViewHelper`, layout/animation) is reusable as-is. The spec now documents three options to explore during development:

- **Configuration-based:** A config struct abstracts the 4 coupling points. Single `TotalsView` for both contexts.
- **Extract sub-views:** Make private sub-views internal, compose a booking payment view from the same parts.
- **Hybrid:** Decouple just the 2 views that read `@Environment(PointOfSaleAggregateModel.self)` to accept closures — may be enough on its own.

Updated in 05e57cb — B1 now covers the coupling analysis and options, B2 is the booking payment controller that provides booking-specific values (e.g. "Done" instead of "New order", no barcode, standard receipt template).

#### B3. POSBookingCashPaymentView
**New file:** `PointOfSale/Presentation/Bookings/POSBookingCashPaymentView.swift`

- Closure-based actions (no `@Environment(PointOfSaleAggregateModel.self)`)
- `FormattableAmountTextField` + `CollectCashViewHelper` for currency/change
- Success state with receipt sending

**Mirrors:** `PointOfSaleCollectCashView.swift` but with closures

#### B4. Receipt handling — standard template for bookings
**Modify:** `PointOfSale/Utils/POSReceiptSender.swift`

Add `isBookingOrder: Bool = false` parameter. When `true`, force `isEligibleForPOSReceipt = false` to use standard receipt template instead of POS template.

---

### Phase 3: Convergence

#### C1. Wire payment into POSBookingsContainerView

> - Create `POSBookingPaymentController` when user taps "Pay by Card" / "Pay by Cash"

Reviewer: ```suggestion
- Create `POSBookingPaymentController` when user taps "Collect Payment"
```

- Remote-only, no optimistic updates (architecture open for later)
- Update mock to match

**Existing code:**
- `BookingsRemote.updateBooking()` at `Networking/Remote/BookingsRemote.swift` — single PUT with optional `attendanceStatus`, `bookingStatus`, `note` params
- Uses the v2 API (see Constraints)

#### M2-F2. POSBookingStatusBadgeView — payment status
**New file:** `PointOfSale/Presentation/Bookings/POSBookingStatusBadgeView.swift`

Color-coded badge for `BookingStatus`:
- `unpaid` → `.posWarningLowest` bg, `.posOnWarningLowest` text
- `paid` / `complete` → `.posInfoLowest` bg, `.posOnInfoLowest` text
- `cancelled` → `.posErrorLowest` bg, `.posOnErrorLowest` text
- `pendingConfirmation` / `confirmed` → `.posDefault` bg, `.posOnDefault` text

Uses `.posCaptionRegular` font, `POSPadding.small`/`.xSmall`, `POSCornerRadiusStyle.small`.

**Mirrors:** `PointOfSale/Presentation/Orders/POSOrderBadgeView.swift`

#### M2-F3. POSBookingAttendanceBadgeView — attendance status
**New file:** `PointOfSale/Presentation/Bookings/POSBookingAttendanceBadgeView.swift`

Color-coded badge for `BookingAttendanceStatus`:
- `booked` → `.posDefault` bg, `.posOnDefault` text

> - `checkedIn` → `.posSuccessLowest` bg, `.posOnSuccessLowest` text

Reviewer: What happens after checked in? Is there a completed status?

Author: Investigation from Claude:

## Booking Statuses vs Attendance Statuses — Two Independent Dimensions

This is an important distinction because these are **two separate, orthogonal status systems** in WooCommerce Bookings:

### Booking Statuses (payment/lifecycle)
Defined in `Networking/Model/Bookings/Booking.swift` as `BookingStatus`:
- `unpaid` — requires payment
- `pendingConfirmation` — awaiting admin approval
- `confirmed` — admin approved, payment still needed
- `paid` — payment received
- **`complete`** — paid, and the booking date is in the past (the service was delivered)
- `cancelled`
- (`in-cart` — internal, used when a customer has the booking in their WooCommerce cart)

### Attendance Statuses (did the customer show up?)
Defined in the same file as `BookingAttendanceStatus`:
- `booked` — appointment is scheduled, hasn't happened yet
- `checkedIn` (`"checked-in"`) — customer arrived and session took place
- `noShow` (`"no-show"`) — customer missed the appointment
- `cancelled`

These two dimensions are **independent**. A booking can be `paid` + `booked`, or `paid` + `checkedIn`, or `complete` + `noShow`, etc.

## What Happens After "Checked-in"?

**"Checked-in" is a terminal attendance status** — there is nothing "after" it in the attendance dimension. The attendance statuses represent a manual assessment by the merchant:
- `booked` → (merchant marks) → `checkedIn` or `noShow`

As described in the `UpdateAttendanceStatusView`:
- **Booked**: "The appointment is scheduled but hasn't happened yet."
- **Checked-in**: "The customer arrived and the session took place as planned."
- **No-show**: "The client missed the appointment without canceling in advance."

The attendance section is hidden entirely when the booking is cancelled (see `BookingDetailsViewModel.setupAttendanceSectionVisibility()`).

## The "Complete" Status — Automatic Transition

Yes, **"complete" exists as a booking status** (not an attendance status). Per the [official WooCommerce Bookings documentation](https://woocommerce.com/document/introduction-to-woocommerce-bookings/managing-woocommerce-bookings/all-bookings/):

> **Complete** — Paid, and the date is in the past. The assumption is that the experience/service was delivered.

The transition from `paid` → `complete` is **automatic and server-side** — WooCommerce Bookings runs a scheduled event (`wc-booking-complete`) that marks paid bookings as complete once their end date has passed. The merchant does not manually trigger this.

## Implications for POS

1. **"Checked-in" does not trigger "complete"** — these are independent systems. Marking attendance as `checkedIn` does not change the booking status. The booking becomes `complete` automatically when (a) it is `paid` and (b) its date has passed.

2. **In the POS flow**, the typical lifecycle would be:
   - Booking arrives as `confirmed`/`unpaid` → merchant collects payment → booking becomes `paid`
   - Merchant marks attendance as `checkedIn` (or `noShow`)
   - Later, the server automatically transitions the booking status from `paid` → `complete`

3. **The `canCancel` check** in the spec correctly excludes `complete` bookings: bookings that are already `complete` (service delivered, past date) should not be cancellable. This is consistent with `BookingDetailsViewModel.isBookingCancellable` which excludes `.cancelled`, `.complete`, and `.unknown`.

4. **No POS action needed for "complete"** — it happens server-side. POS only needs to read and display it properly (which the badge view spec already handles: `paid`/`complete` both get `.posInfoLowest` styling).

- `noShow` → `.posErrorLowest` bg, `.posOnErrorLowest` text
- `cancelled` → muted/grey

**Existing enums:**
- `BookingAttendanceStatus` at `Networking/Model/Bookings/Booking.swift` — booked, checkedIn, cancelled, noShow, unknown
- `BookingStatus` at same file — complete, paid, unpaid, cancelled, pendingConfirmation, confirmed, unknown

---

### Phase 2: Parallel Streams

### Stream A: Detail & List Enrichment

#### M2-A1. POSBooking model enrichment
**Modify:** `PointOfSale/Models/POSBooking.swift`

Add properties:
- `duration: String` (formatted from start/end, e.g. "1h 30m")

> - `timeRange: String` (e.g. "9:00 AM – 10:30 AM")

Reviewer: Can bookings span multiple days? I presume yes...

Author: Investigation from Claude:

## Can bookings span multiple days?

**Yes, bookings can absolutely span multiple days.** The WooCommerce Bookings plugin supports multi-day bookings (e.g. hotel stays, multi-day equipment rentals, multi-day workshops). This is configured at the product level via `duration_unit` (which can be `day`, `hour`, `minute`, or `month`) and `duration_type` (`fixed` for merchant-defined duration, `customer` for customer-chosen duration).

## How dates are represented in the API

Each booking has two timestamp fields:

- **`start`** — Unix timestamp (e.g. `1759417200`) for when the booking begins
- **`end`** — Unix timestamp (e.g. `1759420800`) for when the booking ends

These are decoded as `Date` objects in the iOS `Booking` model (`Booking.swift` lines 122, 129):
```swift
let endDate = Date(timeIntervalSince1970: try container.decode(Double.self, forKey: .endDate))
let startDate = Date(timeIntervalSince1970: try container.decode(Double.self, forKey: .startDate))
```

For a multi-day booking (e.g. a 3-night hotel stay Jan 1-3), `start` would be Jan 1 and `end` would be Jan 3 (or Jan 4 depending on checkout semantics). The `POSBooking` model carries both `startDate` and `endDate` through to the UI layer.

## The `all_day` field

**`all_day`** is a `Bool` field on the `Booking` model (API key: `"all_day"`). When `true`, it indicates the booking spans the entire day without specific start/end times — think "full day rental" rather than "9am-5pm appointment." The `start` and `end` timestamps are still present (typically midnight-to-midnight for the booked day(s)), but the UI should not display specific times for all-day bookings. The test fixture shows `"all_day": false` for time-specific bookings.

Currently, the `allDay` field is decoded and stored in the `Booking` model but is **not mapped through to `POSBooking`** and is **not used in any UI rendering** — `POSBookingRowView` and `POSBookingDetailView` always display a time range regardless of `allDay`. This is a gap that should be addressed (e.g. showing "All Day" instead of a time range when `allDay == true`).

## Implications for "today's bookings" filtering

This is the critical issue. The current filtering uses **`start_date_before` and `start_date_after` API parameters only**, which filter exclusively by `start` date. The `BookingListTab.today` implementation (in `BookingListContainerViewModel.swift`) sets:

```swift
case .today: currentDate.startOfDay(timezone: utcTimeZone) // startDateAfter
case .today: currentDate.endOfDay(timezone: utcTimeZone)   // startDateBefore
```

**The problem:** A multi-day booking that started yesterday (Jan 1) and ends tomorrow (Jan 3) would **not** appear in "today's bookings" (Jan 2) because its `start` date is Jan 1, which falls outside the `start_date_after` filter for today. The API has no `end_date_before`/`end_date_after` filter parameters — there is no way to filter by end date via the REST API.

**The same issue exists in the main app's Bookings tab** — `BookingListViewModel` uses the same `start_date_before`/`start_date_after` filters, so multi-day bookings only appear on the day they start.

**Possible mitigations:**
1. **Widen the date filter** — fetch bookings where `start_date_after` is some days before today and then client-side filter to include bookings where `today >= startDate && today <= endDate`. Downside: over-fetching.
2. **Server-side enhancement** — add `end_date_before`/`end_date_after` filter parameters to the `wc-bookings/v2` API. This would be the cleanest solution.
3. **Accept the limitation** — for M1, document that "today's bookings" means "bookings starting today" and address multi-day overlap in a later iteration.

For POS specifically, option 3 seems reasonable for M1 since the primary use case is appointment-style bookings (hair salon, fitness class) that are typically same-day. Multi-day bookings (hotel stays) are less common in a POS context. But it is worth noting the limitation.

Author: > Accept the limitation — for M1, document that "today's bookings" means "bookings starting today" and address multi-day overlap in a later iteration.

I think this is a reasonable limitation. But I'm clarifying 0fzXPGPRTPyjU3ExPLW6To-fi-37_8520#1624109656

Reviewer: I think the today list should only show bookings starting today, not bookings ongoing today. I can see an argument for also showing bookings _ending_ today but that's less likely to be an issue.

For display, we'll need to make some design decisions... the ranges and dates can get long.

Reviewer 2: Update from design 0fzXPGPRTPyjU3ExPLW6To-fi-151-2582#1624109656

> No multiple-day bookings for MVP. This scenario is more common for events.

| M2-A1 (model), M2-A2 (detail), M2-A3 (row)    | M2-B1 (actions), M2-B2/B3 (action UI)        |
| Testing (badges, computed properties)           | M2-B4 (analytics) + testing (actions)         |

---

## Milestone 3 — Polish & Release (Feb 27)

### Phase 1: Foundation

#### M3-F1. POSBookingFilterState
**New file:** `PointOfSale/Models/POSBookingFilterState.swift`

```swift
struct POSBookingFilterState: Equatable {
    var teamMembers: [BookingTeamMemberFilter]
    var products: [BookingProductFilter]
    var attendanceStatuses: [BookingAttendanceStatus]
    var customers: [BookingCustomerFilter]
    var dateRange: BookingDateRangeFilter?
    var sortOrder: BookingsRemote.Order  // .ascending (oldest first) or .descending (newest first)
}
```

Sort is always by `startDate` — matching the main app's Bookings tab behavior. No `BookingSortOption` enum needed.

Reuses existing data models from the main app Bookings tab:
- `BookingTeamMemberFilter` (name + resourceID)
- `BookingProductFilter` (name + productID)
- `BookingCustomerFilter` (name + customerID)
- `BookingDateRangeFilter` (startDate + endDate)

> - `BookingAttendanceStatus` enum

Reviewer: Explore new data models with claude

Author: ## Investigation: Data Models for POS Bookings — What Exists vs What's Needed

### 1. Existing Models Inventory

#### Networking Layer (`Modules/Sources/Networking/Model/Bookings/`)
| Model | Purpose | Key Fields |
|-------|---------|------------|
| `Booking` | Core API entity, decoded from `wc-bookings/v2` | `bookingID`, `allDay`, `cost`, `customerID`, `startDate`, `endDate`, `orderID`, `productID`, `resourceID`, `statusKey`, `attendanceStatusKey`, `localTimezone`, `currency`, `orderInfo`, `note` |
| `BookingStatus` (enum) | Payment/booking status | `complete`, `paid`, `unpaid`, `cancelled`, `pendingConfirmation`, `confirmed`, `unknown` |
| `BookingAttendanceStatus` (enum) | Attendance tracking | `booked`, `checkedIn`, `cancelled`, `noShow`, `unknown` |
| `BookingResource` | Staff/equipment resource | `resourceID`, `name`, `quantity`, `role`, `email`, `phoneNumber`, `imageID`, `imageURL`, `description` |
| `BookingOrderInfo` | Enrichment from linked order | `statusKey`, `paymentInfo`, `customerInfo`, `productInfo` |
| `BookingPaymentInfo` | Payment details from order | `paymentMethodID`, `paymentMethodTitle`, `subtotal`, `subtotalTax`, `total`, `totalTax` |
| `BookingCustomerInfo` | Customer from order billing | `billingAddress` (Address), `note` |
| `BookingProductInfo` | Product name from order items | `name` |
| `BookingFilters` | Query filters for remote | `productIDs`, `customerIDs`, `resourceIDs`, `startDateBefore/After`, `attendanceStatuses`, `paymentStatuses` |

#### Yosemite Layer (`Modules/Sources/Yosemite/`)
| Model | Purpose | Key Fields |
|-------|---------|------------|
| `POSBooking` | POS-specific display model | `id`, `customerName`, `serviceName`, `startDate`, `endDate`, `formattedAmount`, `status`, `attendanceStatus`, `orderID`, `resourceName` |
| `POSBookingMapper` | Maps `Booking` + `BookingOrderInfo` + `BookingResource` -> `POSBooking` | Uses `CurrencyFormatter` for amount |
| `POSBookingService` | Fetches & enriches bookings | Batch-fetches orders + resources in parallel |
| `POSBookingServiceProtocol` | Service protocol | `fetchBookings(siteID:pageNumber:pageSize:searchQuery:)` |
| `POSBookingListFetchStrategy` | Strategy protocol for list fetching | `fetchBookings(pageNumber:)`, `supportsCaching`, `showsLoadingWithItems` |
| `POSDefaultBookingListFetchStrategy` | Default fetch (no filters) | Delegates to `POSBookingServiceProtocol` |
| `POSSearchBookingListFetchStrategy` | Search fetch | Passes `searchTerm` to service |
| `POSBookingListFetchStrategyFactory` | Creates strategies | `defaultStrategy()`, `searchStrategy(searchTerm:)` |
| `BookingTeamMemberFilter` | Filter by resource/staff | `resourceID`, `name` |
| `BookingProductFilter` | Filter by product | `productID`, `name` |
| `BookingCustomerFilter` | Filter by customer | `customerID`, `name` |
| `BookingDateRangeFilter` | Filter by date range | `startDate`, `endDate` |
| `StoredBookingFilters` | Persisted filters per site | Wraps all filter types |

#### PointOfSale Layer (`Modules/Sources/PointOfSale/`)
| Model | Purpose | Key Fields |
|-------|---------|------------|
| `POSBookingListState` (enum) | View state for booking list | `loading([POSBooking])`, `loaded([POSBooking], hasMoreItems:)`, `inlineError(...)`, `error(...)`, `empty` |
| `POSBookingsModel` | Observable orchestration model | Currently only holds `bookingsController` |
| `POSBookingListController` | List pagination + caching + search | Manages `bookingsViewState`, `selectedBooking`, strategy switching |
| `POSBookingListEmptyViewModel` | Empty state text | Distinct search vs default messages |
| `PointOfSalePaymentState` | Existing POS payment state | `card: PointOfSaleCardPaymentState`, `cash: PointOfSaleCashPaymentState` |

---

### 2. Gaps Identified

#### Gap 1: `POSBooking` is missing M2 enrichment fields
Current `POSBooking` has 10 fields. The spec (M2-A1) calls for these additional properties:
- `duration: String` (formatted, e.g. "1h 30m")
- `timeRange: String` (e.g. "9:00 AM – 10:30 AM")
- `notes: String` (booking notes)
- `customerEmail: String?` (from order billing address)
- `customerPhone: String?` (from order billing address)
- Computed: `canCheckIn`, `canCancel`, `canCollectPayment`

**Data source:** `notes` comes from `Booking.note` (already in API model). `customerEmail`/`customerPhone` come from `BookingOrderInfo.customerInfo.billingAddress` which contains `Address` (has `email` and `phone` fields). Duration and time range are computed from `startDate`/`endDate`. All data is available -- just not yet mapped into `POSBooking`.

#### Gap 2: `POSBookingServiceProtocol` is missing action methods
Current protocol only has `fetchBookings()`. The spec requires:
- `markBookingAsPaid(siteID:bookingID:)` — calls `BookingsRemote.updateBooking(status: .paid)`
- `updateAttendanceStatus(siteID:bookingID:status:)` — calls `BookingsRemote.updateBooking(attendanceStatus:)`
- `cancelBooking(siteID:bookingID:)` — calls `BookingsRemote.updateBooking(bookingStatus: .cancelled)`

**Note:** `BookingsRemote.updateBooking()` already exists and supports all three operations via a single PUT endpoint with optional `attendanceStatus`, `bookingStatus`, and `note` parameters. `BookingStore.markBookingAsPaid()` also exists in the main app layer but uses the Stores/Actions pattern rather than async/await.

#### Gap 3: `POSBookingServiceProtocol.fetchBookings()` doesn't accept `filters` or `order` parameters
The current signature is:
```swift
func fetchBookings(siteID:, pageNumber:, pageSize:, searchQuery:) async throws -> PagedItems<POSBooking>
```
But `BookingsRemote.loadAllBookings()` already accepts `filters: BookingFilters?` and `order: BookingsRemote.Order`. The POS service currently hardcodes `filters: nil` and `order: .ascending`. M3 filtering and sorting will require passing these through.

#### Gap 4: No `POSBookingFilterState` model yet
The spec (M3-F1) plans a new struct:
```swift
struct POSBookingFilterState: Equatable {
    var teamMembers: [BookingTeamMemberFilter]
    var products: [BookingProductFilter]
    var attendanceStatuses: [BookingAttendanceStatus]
    var customers: [BookingCustomerFilter]
    var dateRange: BookingDateRangeFilter?
    var sortOrder: BookingsRemote.Order
}
```
The underlying filter models (`BookingTeamMemberFilter`, etc.) already exist in Yosemite. This is a composition model that needs to be created.

#### Gap 5: No `POSFilteredBookingListFetchStrategy`
M3-F3 needs a third strategy that builds `BookingFilters` from `POSBookingFilterState` and passes them to the service. The factory also needs a `filteredStrategy(state:)` method.

#### Gap 6: `POSBookingsModel` is minimal
Currently only holds a `bookingsController`. Per the spec, it needs to own:
- `POSBookingPaymentController` (M1-B1) — full card/cash payment state machine
- Receipt sender integration with `isBookingOrder: true`
- M2 action methods: `checkIn()`, `markNoShow()`, `cancelBooking()`
- Payment dependencies: `CardPresentPaymentFacade`, order provider

#### Gap 7: No `POSBookingPaymentController`
This is the most significant missing piece for M1. It needs to:
- Manage `PointOfSalePaymentState` (card + cash)
- Fetch the linked order via `orderProvider.fetchOrder()`
- Drive card payment via `CardPresentPaymentFacade`
- Handle `validatingOrder` / `validatingOrderError` states
- Call `markBookingAsPaid()` on success
- Manage subscription-based reader reconnection

**Key difference from POS products:** Error states need configurable button text (e.g. "Back to Booking" instead of "New Order" / "Edit Order").

#### Gap 8: `POSOrder` model is missing `createdVia` field
M3-B2 requires adding `createdVia: String` to distinguish booking orders in the POS order list. The source `Order` model already has this field.

#### Gap 9: No badge view models for POS
M2-F2/F3 need `POSBookingStatusBadgeView` and `POSBookingAttendanceBadgeView`. The main app has `BookingBadgeView` with color mappings, but it uses the main app's color system. POS needs its own using POS design tokens (`posWarningLowest`, `posInfoLowest`, `posErrorLowest`, etc.).

#### Gap 10: `POSBookingDetailView` is missing payment actions
The current detail view is read-only. It displays booking info and status but has no "Collect Payment", "Check In", "No Show", or "Cancel" buttons. These need to be added across M1 (payment) and M2 (actions).

---

### 3. New Models Needed for M2/M3

#### M2 — New Models
1. **`POSBookingStatusBadgeView`** — SwiftUI view with POS color tokens for `BookingStatus` (no new data model, just view)
2. **`POSBookingAttendanceBadgeView`** — SwiftUI view with POS color tokens for `BookingAttendanceStatus` (same)
3. **Enriched `POSBooking`** — Not a new model, but the existing struct needs 5+ new fields and 3 computed properties

#### M3 — New Models
1. **`POSBookingFilterState`** — Composition struct wrapping existing filter models + sort order
2. **`POSFilteredBookingListFetchStrategy`** — New strategy conforming to existing protocol

#### M1 (still in progress) — Missing Models/Controllers
1. **`POSBookingPaymentController`** — Observable class managing payment state machine (the biggest piece)
2. **Expanded `POSBookingServiceProtocol`** — needs `markBookingAsPaid`, and later `updateAttendanceStatus`, `cancelBooking`

---

### 4. Recommendations

1. **POSBooking enrichment (M2-A1) is low-risk and can start early.** All the data already flows through the mapper — it just needs to extract more fields from `BookingOrderInfo.customerInfo.billingAddress` (email, phone) and `Booking.note` (notes). Duration/timeRange are pure computations from existing dates.

2. **`POSBookingServiceProtocol` should be expanded incrementally.** Add `markBookingAsPaid` for M1 payment, then `updateAttendanceStatus`/`cancelBooking` for M2. The underlying `BookingsRemote.updateBooking()` already handles all these via a single PUT.

3. **The `POSBookingPaymentController` is the critical M1 blocker.** It needs to replicate much of `PointOfSaleAggregateModel`'s payment logic but adapted for the booking context (no cart, no order creation, configurable error button text). Consider extracting shared payment subscription patterns into a reusable helper.

4. **For M3 filtering, `POSBookingFilterState` is a thin composition layer** over existing filter models. The main work is building the SwiftUI filter UI to replace the UIKit `FilterListViewController` used in the main app's Bookings tab. The `BookingDateTimeFilterView` is already SwiftUI and may be adaptable.

5. **`POSBookingMapper` needs to evolve in parallel with `POSBooking`.** Each new field added to `POSBooking` requires a corresponding mapping in the mapper. The mapper's signature may need expansion (e.g. to receive raw `Booking` fields like `note` that currently aren't passed through).

6. **No new Networking models are needed.** The API layer (`Booking`, `BookingResource`, `BookingOrderInfo`, `BookingFilters`, `BookingsRemote`) is complete for all three milestones. All gaps are in the Yosemite (service/business logic) and PointOfSale (presentation) layers.

Author: Regarding all this discussion, the fundamental question right now is:

> how much data model detail should the spec capture right now for work that spans M1 through M3?

We could simplify the spec to mention that filtering should support the functionality that Bookings tab support while the UX is still TBD. Then when implementing that part of the functionality we could make the required decisions. @joshheald wdyt?

Reviewer: Sounds good to me.

#### M3-A2. POSBookingListView — add filter button
**Modify:** `PointOfSale/Presentation/Bookings/POSBookingListView.swift`

- Add filter icon button next to search icon in header
- Show active filter count badge
- Present `POSBookingFilterView` as sheet
- On apply: switch strategy to `POSFilteredBookingListFetchStrategy`, reload

#### M3-A3. POSBookingListController — filter strategy support
**Modify:** `PointOfSale/Controllers/POSBookingListController.swift`

- Add `applyFilters(state: POSBookingFilterState)` method
- Switches fetch strategy via factory
- Caches current results before filter (like search does)
- `clearFilters()` restores default strategy

#### M3-A4. View related order — present POSOrdersView
**Modify:** `PointOfSale/Presentation/Bookings/POSBookingDetailView.swift`

- Add "View Order" button when booking has a linked order and order is visible in POS order list
- Present `POSOrdersView` as `.posFullScreenCover` on top of bookings
- `POSOrderListModel` is `@Observable` and propagates through `.fullScreenCover` automatically
- Add `preselectedOrderID` parameter to `POSOrdersView` — after orders load, select matching order instead of first

> - Requires M3-B1 (booking orders in POS order list) so the order is actually in the list

Reviewer: Alternative to explore – push the order in the detail view, don't show the order list.

Author: That could save some development, yes. Absolutely doable.

Reviewer: I tried it on the POC, it works well enough. It might be good to pre-load order details whenever we select a booking.

**Modify:** `PointOfSale/Presentation/Orders/POSOrdersView.swift`
- Add optional `preselectedOrderID: Int64?` parameter
- After orders load, if set, find and select that order

> ### Stream B: Order List Integration

Reviewer: Could we avoid this work by pushing the order from the booking?

Author: We could choose to open Order Details from the booking. The only caveat is that Orders view wouldn't be displaying bookings but maybe that's not a problem for this scope.

Author: Yes! Documented this as the preferred approach in 2b843dbc1b — M3-A4 now describes pushing `POSOrderDetailsView` directly from booking detail instead of presenting the full order list.

`POSOrderListModel` is already in the environment (injected at `PointOfSaleEntryPointView`), and `POSOrderListService.loadOrder(orderID:)` can fetch a single order by ID. Only change needed: expose `loadOrder(orderID:)` on `POSOrderListControllerProtocol` (the fetch strategy is currently private).

This removes the dependency on M3-B1 (order list integration) for the "View Order" feature.

#### M3-B1. Show booking orders in POS order list
**Modify:** `Modules/Sources/NetworkingCore/Remote/OrdersRemote.swift`

Current filter is hardcoded:
```swift
ParameterValues.posFilter = "pos-rest-api"
```

When `.pointOfSaleBookings` is enabled, include booking orders:
- Change `created_via` parameter to array: `["pos-rest-api", "bookings"]`
- WC REST API `created_via` is `type: array` with `compare: IN` — supports this natively
- Apply to both `loadPOSOrders()` and `searchPOSOrders()`

> **Note:** Customer-checkout bookings use `created_via: "checkout"` (classic) or `"store-api"` (block checkout) — these are shared by all WooCommerce orders, not just bookings. Only admin/API-created booking orders (`"bookings"`) should appear. **Needs clarification:** whether customer-created booking orders should also be included and how to identify them.

Reviewer: This is a big issue, raised here https://a8c.slack.com/archives/C09FHQNQERG/p1770401234675609?thread_ts=1770392530.499589&cid=C09FHQNQERG

Author: Yes, I wanted to mention in the spec that i's still under consideration.

Author: Resolved in b070676055 — removed order list integration (M3-B1/B2/B3) from the spec entirely.

Instead of modifying `OrdersRemote`/`POSOrder`/`POSOrderRowView` with `created_via` arrays, we navigate directly to `POSOrderDetailsView` from booking detail via `POSOrderListService.loadOrder(orderID:)`. Single API call, no POS order list changes needed.
